camel/agents/chat_agent.py

```diff
@@ -370,97 +369,75 @@ def step(
-                # Call the model again with return_json_response in the format
-                # of OpenAIFunction as the last tool, returning a structured
-                # response with the user-specified output schema.
-                if output_schema is not None and all(
-                    record.func_name
-                    != Constants.FUNC_NAME_FOR_STRUCTURE_OUTPUT
-                    for record in tool_calls
-                ):
-                    self._add_output_schema_to_tool_list(output_schema)
-
-                    (
-                        response,
-                        output_messages,
-                        finish_reasons,
-                        usage_dict,
-                        response_id,
-                    ) = self._step_model_response(openai_messages, num_tokens)
-
-                    if isinstance(response, ChatCompletion):
-                        # Tools added for function calling and not in stream
-                        # mode
-                        tool_calls, func_assistant_msg, func_result_msg = (
-                            self._add_tools_for_func_call(response, tool_calls)
-                        )
-
-                        # Update the messages
-                        self.update_memory(
-                            func_assistant_msg, OpenAIBackendRole.ASSISTANT
-                        )
-                        self.update_memory(
-                            func_result_msg, OpenAIBackendRole.FUNCTION
-                        )
+                break
 
+            # Check for external tool call
+            tool_call_request = response.choices[0].message.tool_calls[0]
+            if tool_call_request.function.name in self.external_tool_names:
+                # if model calls an external tool, directly return the request
                 info = self._step_get_info(
                     output_messages,
                     finish_reasons,
                     usage_dict,
                     response_id,
-                    tool_calls,
+                    tool_call_records,
                     num_tokens,
+                    tool_call_request,
+                )
+                return ChatAgentResponse(
+                    msgs=output_messages, terminated=self.terminated, info=info
                 )
-                break
 
-        # if use structure response, set structure result as content of
-        # BaseMessage
-        if output_schema and self.model_type.is_openai:
-            for base_message_item in output_messages:
-                base_message_item.content = str(info['tool_calls'][-1].result)
+            # Normal function calling
+            tool_call_records.append(self._step_tool_call_and_update(response))
 
-        chat_agent_response = ChatAgentResponse(
-            msgs=output_messages, terminated=self.terminated, info=info
+        if output_schema is not None and (
+            self.model_type.is_openai or self.model_type.is_gemini
+        ):
```

Comment:
we can add one more property in enum part, like .support_tool_calling, then add openai, gemini, mistral to this

Reply:
This is not updated, please resolve the conversation carefully